controllers/csi/gc/logs.go

```diff
@@ -0,0 +1,103 @@
+package csigc
+
+import (
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/spf13/afero"
+)
+
+const (
+	maxLogFolderSizeBytes = 300000
+	maxNumberOfLogFiles   = 1000
+	maxLogAge             = 14 * 24 * time.Hour
+)
+
+type logFileInfo struct {
+	UnusedVolumeIDs []os.FileInfo
+	NumberOfFiles   int64
+	OverallSize     int64
+}
+
+func (gc *CSIGarbageCollector) runLogGarbageCollection(tenantUUID string) {
+	logs, err := gc.getLogFileInfo(tenantUUID)
+	if err != nil {
+		gc.logger.Info("failed to get log file information")
+		return
+	}
+
+	gc.removeLogsIfNecessary(logs, maxLogFolderSizeBytes, maxNumberOfLogFiles, tenantUUID)
+}
+
+func (gc *CSIGarbageCollector) removeLogsIfNecessary(logs *logFileInfo, maxSize int64, maxFile int64, tenantUUID string) {
+	shouldDelete := logs.NumberOfFiles > 0 && (logs.OverallSize > maxSize || logs.NumberOfFiles > maxFile)
+	if shouldDelete {
+		gc.tryRemoveLogFolders(logs.UnusedVolumeIDs, tenantUUID)
+	}
+}
+
+func (gc *CSIGarbageCollector) getLogFileInfo(tenantUUID string) (*logFileInfo, error) {
+	unusedVolumeIDs, err := gc.getUnusedVolumeIDs(tenantUUID)
+	if err != nil {
+		return nil, err
+	}
+
+	var nrOfFiles int64
+	var size int64
+	for _, volumeID := range unusedVolumeIDs {
+		_ = afero.Walk(gc.fs, filepath.Join(gc.opts.RootDir, tenantUUID, "run", volumeID.Name(), "var"), func(_ string, file os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			if !file.IsDir() {
+				nrOfFiles++
+				size += file.Size()
+			}
+			return nil
+		})
+	}
+
+	return &logFileInfo{
+		UnusedVolumeIDs: unusedVolumeIDs,
+		NumberOfFiles:   nrOfFiles,
+		OverallSize:     size,
+	}, nil
+}
+
+func (gc *CSIGarbageCollector) getUnusedVolumeIDs(tenantUUID string) ([]os.FileInfo, error) {
+	var unusedVolumeIDs []os.FileInfo
+
+	agentDirectoryForPod := filepath.Join(gc.opts.RootDir, tenantUUID, "run")
```

Comment:
This directory is the root for all operations for the logs cleanup and it gets created three times throughout this file. Instead of passing `tenantUUID`, this root directory could be created once in `runLogGarbageCollection` and passed when needed.

Reply:
Moved